core/src/main/java/org/apache/iceberg/io/BaseEqualityDeltaWriter.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.iceberg.util.StructProjection;
+
+public class BaseEqualityDeltaWriter<T> implements EqualityDeltaWriter<T> {
+
+  private final ThreadLocal<PositionDelete<T>> posDelete = ThreadLocal.withInitial(PositionDelete::create);
+
+  private final PartitioningWriter<T, DataWriteResult> dataWriter;
+  private final PartitioningWriter<T, DeleteWriteResult> equalityWriter;
+  private final PartitioningWriter<PositionDelete<T>, DeleteWriteResult> positionWriter;
+
+  private final Map<StructLike, PathOffset> insertedRowMap;
+
+  private final Function<T, StructLike> asStructLike;
+  private final Function<T, StructLike> keyRefFunc;
+  private final Function<T, StructLike> keyCopyFunc;
+
+  private boolean closed = false;
+
+  public BaseEqualityDeltaWriter(
+      PartitioningWriter<T, DataWriteResult> dataWriter,
+      PartitioningWriter<T, DeleteWriteResult> equalityWriter,
+      PartitioningWriter<PositionDelete<T>, DeleteWriteResult> positionWriter,
+      Schema schema,
+      Schema deleteSchema,
+      Function<T, StructLike> asStructLike) {
+    this.dataWriter = dataWriter;
+    this.equalityWriter = equalityWriter;
+    this.positionWriter = positionWriter;
+
+    this.insertedRowMap = StructLikeMap.create(deleteSchema.asStruct());
+    this.asStructLike = asStructLike;
+
+    StructProjection projection = StructProjection.create(schema, deleteSchema);
+    this.keyRefFunc = row -> projection.wrap(asStructLike.apply(row));
+    this.keyCopyFunc = row -> StructCopy.copy(keyRefFunc.apply(row));
+  }
+
+  @Override
+  public void insert(T row, PartitionSpec spec, StructLike partition) {
+    PathOffset pathOffset = dataWriter.write(row, spec, partition);
+
+    // Create a copied key from this row.
+    StructLike copiedKey = keyCopyFunc.apply(row);
+
+    // Adding a pos-delete to replace the old path-offset.
+    PathOffset previous = insertedRowMap.put(copiedKey, pathOffset);
+    if (previous != null) {
+      positionWriter.write(posDelete.get().set(previous.path(), previous.rowOffset(), null), spec, partition);
+    }
+  }
+
+  /**
+   * Retire the old key & position from insertedRowMap cache to position delete file.
+   */
+  private boolean retireOldKey(StructLike key, PartitionSpec spec, StructLike partition) {
+    PathOffset previous = insertedRowMap.remove(key);
+    if (previous != null) {
+      positionWriter.write(posDelete.get().set(previous.path(), previous.rowOffset(), null), spec, partition);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public void delete(T row, PartitionSpec spec, StructLike partition) {
+    if (!retireOldKey(keyRefFunc.apply(row), spec, partition)) {
+      equalityWriter.write(row, spec, partition);
+    }
+  }
+
+  @Override
+  public void deleteKey(T key, PartitionSpec spec, StructLike partition) {
+    if (!retireOldKey(asStructLike.apply(key), spec, partition)) {
+      equalityWriter.write(key, spec, partition);
+    }
```

Comment:
I think this will be needed to reconsider because of this PR: https://github.com/apache/iceberg/pull/4364